src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.damagesource;
+
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.function.BiFunction;
+import net.minecraft.world.damagesource.DamageSource;
+import net.neoforged.neoforge.event.entity.living.DamageBlockEvent;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * DamageContainer encapsulates aspects of the entity damage sequence so that
+ * relevant context related to damage dealt is accessible throughout the entire
+ * sequence.
+ * <h4>Note: certain values will be defaults until the stage in the sequence when they are set.</h4>
+ * <H3>Damage Sequence and uses</H3><ol>
+ * <li>Entity is hurt by a damage source</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.EntityInvulnerablityCheckEvent EntityInvulnerablityCheckEvent}
+ * fires and determines if the sequence can commence</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.EntityPreDamageEvent EntityPreDamageEvent} fires
+ * and gives access to this. Modifiers should be added here.</li>
+ * <li>{@link DamageBlockEvent} fires</li>
+ * <li>armor, enchantments, mob effect, and absorption modifiers are applied to the damage</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent IncomingDamageEvent} fires and
+ * provides final values for the preceding modifiers and the last chance to negate the damage but will not
+ * undo the preceding effects</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.DamageTakenEvent DamageTakenEvent} fires and provides
+ * an immutable perspective of what the entire sequence ended with. </li>
+ * </ol>
```

Comment:
Javadoc references should typically use imported names (instead of FQNs) to reduce verboseness outside of an IDE.